src/gui/base/OfflineIndicatorViewModel.ts

```diff
@@ -0,0 +1,121 @@
+import {PROGRESS_DONE} from "./ProgressBar.js"
+import Stream from "mithril/stream"
+import {WorkerClient, WsConnectionState} from "../../api/main/WorkerClient.js"
+import {ExposedCacheStorage} from "../../api/worker/rest/EntityRestCache.js"
+import {ILoginListener} from "../../api/main/LoginListener.js"
+import {LoginController} from "../../api/main/LoginController.js"
+import {OfflineIndicatorAttrs, OfflineIndicatorState} from "./OfflineIndicator.js"
+
+/**
+ * the offline indicator must take into account information
+ * from multiple different sources:
+ * * ws connection state (connected, not connected) from the worker
+ * * login state (logged out, partial login, full login)
+ * * sync progress
+ * * last sync time
+ *
+ * the state necessary to determine the right indicator state from
+ * previous updates from these information sources
+ * is maintained in this class
+ */
+export class OfflineIndicatorViewModel {
+
+	private lastProgress: number = PROGRESS_DONE
+	private lastWsState: WsConnectionState = WsConnectionState.connecting
+	private lastUpdate: Date | null = null
+	private wsWasConnectedBefore: boolean = false
```

Comment:
Why is previous WS important? Isn't it almost the same as fully logged in?

Reply:
there seems to be a short time between "fully logged in" and "websocket connected the first time". The indicator would show the reconnect button in that time if we didn't keep this around.